Geoserver: Install Kafka DS plugin

gdooper opened the pull request:
Installs the Geomesa Kafka DS plugin in Geoserver.

Also refactors the docker build process to use docker build stages for fewer layers and a smaller overall image size.

By default, this will now create a new geoserver image with a tag that includes the Kafka version. E.g.:
`quay.io/geomesa/geoserver:geomesa-<geomesa.version>-accumulo-<accumulo.version>-kafka-<kafka.version>`

gdooper committed:
Install Kafka DS Geoserver plugin

gdooper committed:
Refactor using build stages for a smaller image

jnh5y commented on geodocker-geoserver/Dockerfile.template:
```diff
@@ -1,22 +1,28 @@
+  | tar -zx
 
-ARG TAG
+FROM centos:7
```
I don't Docker much; does this 'FROM' directive need to be repeated here?

jahhulbert-ccri replied:
should only be at top

gdooper replied:
These multiple FROM statements are used for multi-stage builds in Docker (https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#use-multi-stage-builds).
I did this to help reduce the number of layers in the final image and therefore reduce the final image size. It used to be ~2GB, but with multi-stage builds it's down to ~1GB

jahhulbert-ccri replied:
whaaat version 17.05 man I'm old and behind. This is cool.

jnh5y commented:
@jahhulbert-ccri @anthonyccri thoughts on this?

jahhulbert-ccri commented on geodocker-geoserver/Makefile:
```diff
@@ -2,36 +2,31 @@ BASE := $(subst -, ,$(notdir ${CURDIR}))
-GEOMESA_DIST_TARBALL := geomesa-accumulo-dist_2.11-${GEOMESA_VERSION}-bin.tar.gz
-GEOMESA_ACCUMULO_GS_INSTALL := geomesa-accumulo-gs-plugin_2.11-${GEOMESA_VERSION}-install.tar.gz
+GEOMESA_VERSION := 2.0.0-m.1
+ACCUMULO_VERSION := 1.7.3
```
use accumulo 1.8.1 and 0.9.3 thrift by default to avoid commons-vfs issues

gdooper replied:
I made the ACCUMULO version and THRIFT version consistent with the default version found in [geodocker-geomesa/build.sh](https://github.com/geodocker/geodocker-geomesa/blob/18177fd402230bfaabd83c14b3d05cbb33846fa6/build.sh#L5).

If we want to default ACCUMULO_VERSION to 1.8.1, I think it would be best to do that in a separate PR and across this entire project

jahhulbert-ccri replied:
sounds good. I'd recommend that due to issues with 1.7.3 (1.7.4 just came out and might have helped). but just go to 1.8.1 i'd recommend.

jahhulbert-ccri commented on geodocker-geoserver/Dockerfile.template:
```diff
@@ -75,7 +75,37 @@ RUN set -x \
+ENV KAFKA_VERSION ${KAFKA_VERSION}
+ENV THRIFT_VERSION ${THRIFT_VERSION}
+ENV HADOOP_VERSION 2.7.3
+ENV GEOSERVER_VERSION 2.9.4
```
this needs to be parameterized in tandem with geomesa...

jahhulbert-ccri replied:
or be bumped if this docker is hard coding version 2.0 geomesa...should be 2.12.x

jahhulbert-ccri commented on geodocker-geoserver/Dockerfile.template:
```diff
@@ -75,7 +75,37 @@ RUN set -x \
+RUN set -x \
+  && cd /opt/tomcat/webapps/geoserver/WEB-INF/lib/ \
+  && tar zxvf /opt/geomesa-kafka/dist/gs-plugins/geomesa-kafka-${KAFKA_VERSION}-gs-plugin_2.11-${GEOMESA_VERSION}-install.tar.gz \
+  && rm /opt/tomcat/webapps/geoserver/WEB-INF/lib/zookeeper*.jar \
```
is there already a ZK jar there?

jahhulbert-ccri replied:
FYI if we can we should use ZK 3.4.10 instead of 3.4.6 everywhere due to bug fixes that have caused accumulo and kafka issues

gdooper replied:
Yes, there already was a zookeeper jar there. When I run the `install-kafka.sh` script to add further dependencies, we ended up with multiple zookeeper jars of differing versions. That's why I deleted the zookeeper jar before running the script.

There's probably a better way to handle this (maybe handling this inside the `install-kafka.sh` script?), but this seemed to work

jahhulbert-ccri replied:
hahaha prob always a better way but just wondering...might be worth a comment to mention it.

gdooper replied:
Hmm, the `install-hadoop-accumulo.sh` script that gets run when installing the accumulo plugin installs zookeeper-3.4.9.jar, but `install-kafka.sh` script installs zookeeper-3.4.5.jar.

Maybe it's worth installing kafka plugin first, delete zookeeper-3.4.5.jar, then install accumulo plugin with zookeeper-3.4.9.jar

gdooper committed:
Merge branch 'master' into geoserver-kafka-plugin

 Conflicts:
	geodocker-geoserver/Dockerfile.template

gdooper committed:
Update geoserver/tomcat versions

gdooper commented:
@jnh5y @jahhulbert-ccri I've made a couple requested changes and added some commits. Have you had a chance to look again?

jnh5y commented:
@jahhulbert-ccri can you review this again?

elahrvivaz commented:
pinging @jahhulbert-ccri

elahrvivaz commented:
@gdooper sorry I forgot about this - if you could fix the merge conflicts, then I will go ahead and merge it. @jahhulbert-ccri last chance to chime in

gdooper committed:
Merge branch 'master' into geoserver-kafka-plugin

Conflicts:
	build.sh
	geodocker-geoserver/Dockerfile.template

gdooper committed:
Update HADOOP_VERSION env

gdooper committed:
Remove KAFKA_VERSION references

Seems like geomesa v2.1.0 doesn't include kafka
version in build artifacts. This will likely
break compatibility with earlier geomesa versions

gdooper committed:
Build geoserver with locally installed geomesa distributions

gdooper commented:
@elahrvivaz I've resolved conflicts and updated to work with geomesa 2.1.0. This might break the build when trying to use with an older version of geomesa that includes the kafka version in the geomesa-kafka artifact names.

I also reverted back to using a locally installed version of the geomesa-accumulo and geomesa-kafka distributions, just so you don't have to download it from the locationtech repo.

elahrvivaz commented:
@gdooper thanks!
